### PR TITLE
[Draft] SSO Protocol

### PR DESCRIPTION
**Pull request by Jeffders**
Fixes https://github.com/microsoft/botbuilder-dotnet/issues/3356

### CODE REVIEW

File: libraries/Microsoft.Bot.Builder.AI.QnA/Utils/GenerateAnswerUtils.cs [hunk 1]
```diff
@@ -143,6 +143,11 @@ private static void ValidateOptions(QnAMakerOptions options)
             {
                 options.MetadataBoost = new Metadata[] { };
             }
+
```
Comment (carlosscastro) [inline libraries/Microsoft.Bot.Builder.AI.QnA/Utils/GenerateAnswerUtils.cs]: Where did this come from?

File: libraries/Microsoft.Bot.Builder.Dialogs/Extensions.cs [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    /// </summary>
+    internal static partial class Extensions
+    {
+        /// <summary>
+        /// Extension Method on object to cast to type T to support TypeNameHandling.None during storage serialization.
+        /// </summary>
+        /// <param name="obj">object to cast.</param>
+        /// <typeparam name="T">type to which object should be casted.</typeparam>
+        /// <returns>T.</returns>
+        public static T CastTo<T>(this object obj)
```
Comment (carlosscastro) [inline libraries/Microsoft.Bot.Builder.Dialogs/Extensions.cs]: Should add unit tests to cover this extension

File: libraries/Microsoft.Bot.Builder.Dialogs/Prompts/OAuthPrompt.cs [hunk 3]
```diff
@@ -426,12 +433,103 @@ private async Task<PromptRecognizerResult<TokenResponse>> RecognizeTokenAsync(IT
                     }
                     else
                     {
-                        await turnContext.SendActivityAsync(new Activity { Type = ActivityTypesEx.InvokeResponse, Value = new InvokeResponse { Status = 404 } }, cancellationToken).ConfigureAwait(false);
+                        await this.SendInvokeResponseAsync(turnContext, cancellationToken, 404).ConfigureAwait(false);
```
Comment (carlosscastro) [inline libraries/Microsoft.Bot.Builder.Dialogs/Prompts/OAuthPrompt.cs]: Use constant, HttpStatusCodes.NotFound already defined in the framework IIRC (name may vary slightly)

File: libraries/Microsoft.Bot.Builder.Dialogs/Prompts/OAuthPrompt.cs [hunk 4]
```diff
@@ -426,12 +433,103 @@ private async Task<PromptRecognizerResult<TokenResponse>> RecognizeTokenAsync(IT
...
                     else
                     {
-                        await turnContext.SendActivityAsync(new Activity { Type = ActivityTypesEx.InvokeResponse, Value = new InvokeResponse { Status = 404 } }, cancellationToken).ConfigureAwait(false);
+                        await this.SendInvokeResponseAsync(turnContext, cancellationToken, 404).ConfigureAwait(false);
                     }
                 }
                 catch
                 {
-                    await turnContext.SendActivityAsync(new Activity { Type = ActivityTypesEx.InvokeResponse, Value = new InvokeResponse { Status = 500 } }, cancellationToken).ConfigureAwait(false);
+                    await this.SendInvokeResponseAsync(turnContext, cancellationToken, 500).ConfigureAwait(false);
```
Comment (carlosscastro) [inline libraries/Microsoft.Bot.Builder.Dialogs/Prompts/OAuthPrompt.cs]: use constant, HttpStatusCodes.InternalServerError or similar

File: libraries/Microsoft.Bot.Builder.Dialogs/Prompts/OAuthPrompt.cs [hunk 5]
```diff
@@ -426,12 +433,103 @@ private async Task<PromptRecognizerResult<TokenResponse>> RecognizeTokenAsync(IT
...
+            else if (IsTokenExchangeRequestInvoke(turnContext))
+            {
+                var tokenExchangeRequest = ((JObject)turnContext.Activity.Value).ToObject<TokenExchangeInvokeRequest>();
+
+                if (tokenExchangeRequest == null)
+                {
+                    await this.SendInvokeResponseAsync(
+                        turnContext,
+                        cancellationToken,
+                        400,
```
Comment (carlosscastro) [inline libraries/Microsoft.Bot.Builder.Dialogs/Prompts/OAuthPrompt.cs]: Constant from HttpStatusCodes

File: libraries/Microsoft.Bot.Builder.Dialogs/Prompts/OAuthPrompt.cs [hunk 6]
```diff
@@ -426,12 +433,103 @@ private async Task<PromptRecognizerResult<TokenResponse>> RecognizeTokenAsync(IT
...
+                        new TokenExchangeInvokeResponse()
+                        {
+                            Id = tokenExchangeRequest.Id,
+                            ConnectionName = _settings.ConnectionName,
+                            FailureDetail = "Missing TokenExchangeInvokeRequest",
```
Comment (carlosscastro) [inline libraries/Microsoft.Bot.Builder.Dialogs/Prompts/OAuthPrompt.cs]: If a bot developer gets this error, can they take action? If possible I prefer actionable errors, i.e. 'X is not available, please do Y.'

File: libraries/Microsoft.Bot.Builder.Dialogs/Prompts/OAuthPrompt.cs [hunk 7]
```diff
@@ -426,12 +433,103 @@ private async Task<PromptRecognizerResult<TokenResponse>> RecognizeTokenAsync(IT
...
+                        }).ConfigureAwait(false);
+                }
+                else if (tokenExchangeRequest.ConnectionName != _settings.ConnectionName)
+                {
+                    await this.SendInvokeResponseAsync(
+                        turnContext,
+                        cancellationToken,
+                        400,
```
Comment (carlosscastro) [inline libraries/Microsoft.Bot.Builder.Dialogs/Prompts/OAuthPrompt.cs]: Constant: HttpStatusCodes.blah

File: libraries/Microsoft.Bot.Builder.Dialogs/Prompts/OAuthPrompt.cs [hunk 8]
```diff
@@ -426,12 +433,103 @@ private async Task<PromptRecognizerResult<TokenResponse>> RecognizeTokenAsync(IT
...
+                            ConnectionName = _settings.ConnectionName,
+                            FailureDetail = "Invalid ConnectionName in the TokenExchangeInvokeRequest",
+                        }).ConfigureAwait(false);
+                }
+                else if (!(turnContext.Adapter is ITokenExchangeProvider adapter))
+                {
+                    await this.SendInvokeResponseAsync(
+                           turnContext,
+                           cancellationToken,
+                           502,
```
Comment (carlosscastro) [inline libraries/Microsoft.Bot.Builder.Dialogs/Prompts/OAuthPrompt.cs]: Constant: HttpStatusCodes.BadGateway

File: libraries/Microsoft.Bot.Builder.Dialogs/Prompts/OAuthPrompt.cs [hunk 9]
```diff
@@ -426,12 +433,103 @@ private async Task<PromptRecognizerResult<TokenResponse>> RecognizeTokenAsync(IT
...
+                           new TokenExchangeInvokeResponse()
+                           {
+                               Id = tokenExchangeRequest.Id,
+                               ConnectionName = _settings.ConnectionName,
+                               FailureDetail = "Bot Adapter does not support token exchange.",
```
Comment (carlosscastro) [inline libraries/Microsoft.Bot.Builder.Dialogs/Prompts/OAuthPrompt.cs]: Error message: Could recommend how to fix it? 'Bot adapter does not support token exchange. Upgrade your bot to the latest version of the Bot Framework SDK.' Or something that helps users indicating what to do?

File: libraries/Microsoft.Bot.Builder.Dialogs/Prompts/OAuthPrompt.cs [hunk 10]
```diff
@@ -426,12 +433,103 @@ private async Task<PromptRecognizerResult<TokenResponse>> RecognizeTokenAsync(IT
...
+                            Token = tokenExchangeRequest.Token,
+                        },
+                        cancellationToken).ConfigureAwait(false);
+
+                    if (tokenExchangeResponse == null || string.IsNullOrEmpty(tokenExchangeResponse.Token))
+                    {
+                        await this.SendInvokeResponseAsync(
+                           turnContext,
+                           cancellationToken,
+                           409,
```
Comment (carlosscastro) [inline libraries/Microsoft.Bot.Builder.Dialogs/Prompts/OAuthPrompt.cs]: constant

File: libraries/Microsoft.Bot.Builder/BotFrameworkAdapter.cs [hunk 11]
```diff
@@ -894,6 +895,114 @@ public virtual async Task<TokenStatus[]> GetTokenStatusAsync(ITurnContext contex
...
+            }
+
+            if (result is TokenResponse tokenResponse)
+            {
+                // todo: temporary
+                if (string.IsNullOrEmpty(tokenResponse.Token))
+                {
+                    if (!string.IsNullOrEmpty(exchangeRequest.Uri))
+                    {
+                        tokenResponse.Token = "exchangeableToken" + Guid.NewGuid().ToString().Replace("-", string.Empty);
```
Comment (carlosscastro) [inline libraries/Microsoft.Bot.Builder/BotFrameworkAdapter.cs]: Guid.NewGuid().ToString("N");

File: libraries/Microsoft.Bot.Builder/BotFrameworkAdapter.cs [hunk 12]
```diff
@@ -894,6 +895,114 @@ public virtual async Task<TokenStatus[]> GetTokenStatusAsync(ITurnContext contex
...
+                    }
+                    else
+                    {
+                        tokenResponse.Token = "token" + Guid.NewGuid().ToString().Replace("-", string.Empty);
```
Comment (carlosscastro) [inline libraries/Microsoft.Bot.Builder/BotFrameworkAdapter.cs]: Guid.NewGuid().ToString("N");

File: libraries/Microsoft.Bot.Builder/BotFrameworkAdapter.cs [hunk 13]
```diff
@@ -894,6 +895,114 @@ public virtual async Task<TokenStatus[]> GetTokenStatusAsync(ITurnContext contex
...
+            var result = await client.UserToken.ExchangeAsyncAsync(userId, connectionName, turnContext.Activity.ChannelId, exchangeRequest, cancellationToken).ConfigureAwait(false);
+
+            if (result is ErrorResponse errorResponse)
+            {
+                result = new TokenResponse();
+            }
+
+            if (result is TokenResponse tokenResponse)
+            {
+                // todo: temporary
```
Comment (carlosscastro) [inline libraries/Microsoft.Bot.Builder/BotFrameworkAdapter.cs]: What is the solution to this TODO , considering we are trying to get this completed?

File: libraries/Microsoft.Bot.Builder/BotFrameworkAdapter.cs [hunk 14]
```diff
@@ -894,6 +895,114 @@ public virtual async Task<TokenStatus[]> GetTokenStatusAsync(ITurnContext contex
...
+        /// <param name="exchangeRequest">The exchange request details, either a token to exchange or a uri to exchange.</param>
+        /// <param name="cancellationToken">A cancellation token that can be used by other objects
+        /// or threads to receive notice of cancellation.</param>
+        /// <returns>If the task completes, the exchanged token is returned.</returns>
+        public virtual async Task<TokenResponse> ExchangeTokenAsync(ITurnContext turnContext, string connectionName, string userId, TokenExchangeRequest exchangeRequest, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            var client = await CreateOAuthApiClientAsync(turnContext).ConfigureAwait(false);
+            var result = await client.UserToken.ExchangeAsyncAsync(userId, connectionName, turnContext.Activity.ChannelId, exchangeRequest, cancellationToken).ConfigureAwait(false);
+
+            if (result is ErrorResponse errorResponse)
```
Comment (carlosscastro) [inline libraries/Microsoft.Bot.Builder/BotFrameworkAdapter.cs]: Could you explain hwat the logic is here? SHouldn't we log the error and that we are falling back to something else?

File: libraries/Microsoft.Bot.Builder/BotFrameworkAdapter.cs [hunk 15]
```diff
@@ -894,6 +895,114 @@ public virtual async Task<TokenStatus[]> GetTokenStatusAsync(ITurnContext contex
...
+                        tokenResponse.Token = "exchangeableToken" + Guid.NewGuid().ToString().Replace("-", string.Empty);
+                    }
+                    else
+                    {
+                        tokenResponse.Token = "token" + Guid.NewGuid().ToString().Replace("-", string.Empty);
+                    }
+                }
+
+                return tokenResponse;
+            }
```
Comment (carlosscastro) [inline libraries/Microsoft.Bot.Builder/BotFrameworkAdapter.cs]: Food for thought: I already worry about these methods being part of the Bot Framework Adapter. However, now we are getting to the point where a significant % of the adapter code is signin related. I wonder if we should 1) refactor the signin code out to a class, and have the adapter expose the key public methods and call the satellite class, or 2) refactor the signin code to a class, have it implement IUserTokenProvider + ITokenExchangeProvider, and have a property in adapter that exposes its ITokenExchangeProvider, and finally OAuthPrompt consumes this class. That way we can clean up the adapter. Just for thought.

File: libraries/Microsoft.Bot.Builder/ITokenExchangeProvider.cs [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+        /// </summary>
+        /// <param name="turnContext">Context for the current turn of conversation with the user.</param>
+        /// <param name="connectionName">Name of the auth connection to use.</param>
+        /// <param name="userId">The user id that will be associated with the token.</param>
+        /// <param name="finalRedirect">The final URL that the OAuth flow will redirect to.</param>
+        /// <param name="cancellationToken">A cancellation token that can be used by other objects
+        /// or threads to receive notice of cancellation.</param>
+        /// <returns>A task that represents the work queued to execute.</returns>
+        /// <remarks>If the task completes successfully, the result contains the raw signin link.</remarks>
+        Task<SignInResource> GetSignInResourceAsync(ITurnContext turnContext, string connectionName, string userId, string finalRedirect = null, CancellationToken cancellationToken = default(CancellationToken));
```
Comment (carlosscastro) [inline libraries/Microsoft.Bot.Builder/ITokenExchangeProvider.cs]: Remember that this interface cannot be changed in the future. What could change here? What could we need in the future? Let's play 'change storm' and think of what else we could need here

File: libraries/Microsoft.Bot.Builder.AI.QnA/Utils/GenerateAnswerUtils.cs (hunk 1 above)
Comment (Jeffders) [reply libraries/Microsoft.Bot.Builder.AI.QnA/Utils/GenerateAnswerUtils.cs]: no idea, it's in the 4.7 branch...

---
In reply to: [375486469](https://github.com/microsoft/botbuilder-dotnet/pull/3320#discussion_r375486469) [](ancestors = 375486469)

File: libraries/Microsoft.Bot.Builder.Dialogs/Extensions.cs (hunk 2 above)
Comment (Jeffders) [reply libraries/Microsoft.Bot.Builder.Dialogs/Extensions.cs]: I didn't add this...it's just in 4.7 branch :p

---
In reply to: [375486985](https://github.com/microsoft/botbuilder-dotnet/pull/3320#discussion_r375486985) [](ancestors = 375486985)

File: libraries/Microsoft.Bot.Builder/BotFrameworkAdapter.cs (hunk 12 above)
Comment (Jeffders) [reply libraries/Microsoft.Bot.Builder/BotFrameworkAdapter.cs]: This code should be removed.

---
In reply to: [375491094](https://github.com/microsoft/botbuilder-dotnet/pull/3320#discussion_r375491094) [](ancestors = 375491094)

File: libraries/Microsoft.Bot.Builder/BotFrameworkAdapter.cs (hunk 11 above)
Comment (Jeffders) [reply libraries/Microsoft.Bot.Builder/BotFrameworkAdapter.cs]: This code should be removed.

---
In reply to: [375491057](https://github.com/microsoft/botbuilder-dotnet/pull/3320#discussion_r375491057) [](ancestors = 375491057)

File: libraries/Microsoft.Bot.Builder/BotFrameworkAdapter.cs (hunk 15 above)
Comment (Jeffders) [reply libraries/Microsoft.Bot.Builder/BotFrameworkAdapter.cs]: I like the second option...can we really do it? isn't it "brekaing" at this point?

---
In reply to: [375493100](https://github.com/microsoft/botbuilder-dotnet/pull/3320#discussion_r375493100) [](ancestors = 375493100)

File: libraries/Microsoft.Bot.Builder/BotFrameworkAdapter.cs (hunk 14 above)
Comment (Jeffders) [reply libraries/Microsoft.Bot.Builder/BotFrameworkAdapter.cs]: no, it should throw :p

---
In reply to: [375491544](https://github.com/microsoft/botbuilder-dotnet/pull/3320#discussion_r375491544) [](ancestors = 375491544)

File: libraries/Microsoft.Bot.Builder/BotFrameworkAdapter.cs (hunk 13 above)
Comment (Jeffders) [reply libraries/Microsoft.Bot.Builder/BotFrameworkAdapter.cs]: removed it

---
In reply to: [375491290](https://github.com/microsoft/botbuilder-dotnet/pull/3320#discussion_r375491290) [](ancestors = 375491290)

**Comment by Jeffders** [comment]
Draft complete, abandoning this to create a master-branch mergeable PR.

**Pull request closed without merging**